src/lib/warn-upload-overwrite.js

```diff
@@ -15,139 +16,195 @@ const responseChoicesWithoutDiff = [
+    fs.mkdir(bookmarksDir);
   }
 
-  // Pull local _rev
-  let localRev;
-  try {
-    localRev = JSON.parse(fs.read(`${environment.pathToProject}/._revs/${localDoc._id}.json`).trim())[getRevsAndHashesDocKey()];
-  } catch (e) {
-    // continue regardless of error
-    log.trace('Trying to fetch local _rev', e);
+  const filePath = path.join(bookmarksDir, 'hashes.json');
+  if (fs.exists(filePath)) {
+    try {
+      return JSON.parse(fs.read(filePath).trim());
+    } catch(e) {
+      log.info('Error trying to read bookmark, continuing anyway', e);
+    }
   }
 
-  if (localRev) {
-    localDoc._rev = localRev;
-  }
+  return {};
+};
 
-  if (localDoc._attachments) {
-    for (let key in localDoc._attachments) {
-      localDoc._attachments[key] = {
-        'content_type': doc._attachments[key].content_type, 
-        'revpos': localRev ? parseInt(localRev.split('-')[0]) : 0, 
-        'length': doc._attachments[key].data.length,
-        'digest': `md5-${crypto.createHash('md5').update(doc._attachments[key].data, 'binary').digest('base64')}`,
-        'stub': true
-      };
-    }
-  }
+const getStoredHash = id => {
+  const hashes = getStoredHashes();
+  return hashes && hashes[id] && hashes[id][getEnvironmentKey()];
+};
 
-  let diff; 
-  if (localDoc._id === 'settings' && remoteDoc._id === 'settings') {
-    diff = jsonDiff.diffString(remoteDoc.settings, localDoc.settings);
-  } else {
-    diff = jsonDiff.diffString(remoteDoc, localDoc);
+const updateStoredHash = (id, hash) => {
+  const hashes = getStoredHashes();
+  if (!hashes[id]) {
+    hashes[id] = {};
   }
-  if (diff) {  
-    if (!localRev) {
-      if (!readline.keyInYN(`We can't determine if you're going to overwrite someone else's changes or not. Would you like to proceed?`)) {
-        throw new Error('configuration modified');
-      }
-    } else if (localRev !== remoteRev) {
-      let index = readline.keyInSelect(responseChoicesWithDiff, question, {cancel: false});
-      if (index === 2) { // diff
-        log.info(diff);
+  hashes[id][getEnvironmentKey()] = hash;
+  fs.write(path.join(getBookmarksDir(), 'hashes.json'), JSON.stringify(hashes));
+};
 
-        index = readline.keyInSelect(responseChoicesWithoutDiff, question, {cancel: false});
-      }
-      if (index === 1) { // abort
-        throw new Error('configuration modified');
+const couchDigest = content => 'md5-' + crypto.createHash('md5').update(content, 'binary').digest('base64');
+
+const getXFormAttachment = doc => {
+  const name = Object
+    .keys(doc && doc._attachments || {})
+    .find(name => name === 'xml' || (name.endsWith('.xml') && name !== 'model.xml'));
+  return name;
+};
+
+const getFormHash = (doc, xml) => {
+  const xFormAttachmentName = getXFormAttachment(doc);
+  const crypt = crypto.createHash('md5');
+  crypt.update(xml, 'utf8');
+  const properties = {
+    context: doc.context,
+    icon: doc.icon,
+    internalId: doc.internalId,
+    title: doc.title
```

Comment:
We use the `xml2sms` field to generate our custom SMS format for forms. https://github.com/medic/cht-core/blob/master/webapp/src/js/services/form2sms.js#L23
I think we should include it.

Reply:
Great find! It looks like it's never been possible to configure that with medic-conf. I've raised #306 and will include that in this PR too.

Reply:
Done.